multiapps-controller-core/src/main/java/org/cloudfoundry/multiapps/controller/core/util/NameUtil.java

```diff
@@ -73,19 +61,63 @@ public static String getNameWithProperLength(String name, int maxLength) {
         return name;
     }
 
-    public static String computeNamespacedNameWithLength(String name, String namespace, boolean applyNamespace, int maxLength) {
-        String prefix = "";
+    public static String computeNamespacedNameWithLength(String name, String namespace, boolean applyNamespace,
+                                                         boolean applyNamespaceAsSuffix, int maxLengthWithSuffix) {
         if (StringUtils.isNotEmpty(namespace) && applyNamespace) {
-            prefix = getNamespacePrefix(namespace);
+            if (applyNamespaceAsSuffix) {
+                name = getNameWithNamespaceSuffix(name, namespace, maxLengthWithSuffix);
+            } else {
+                name = getNamespacePrefix(namespace) + name;
+            }
+        }
+        return getNameWithProperLength(name, maxLengthWithSuffix);
+    }
+
+    private static String getNameWithNamespaceSuffix(String name, String namespace, int maxLengthWithSuffix) {
+        String namespaceSuffix = getNamespaceSuffix(namespace);
+        String nameWithoutShortening = name;
+        name = getNameWithProperLength(name, calculateNameLengthWithoutNamespaceAndBlueGreenSuffix(namespaceSuffix, maxLengthWithSuffix));
+
+        name = correctNameSuffix(name, nameWithoutShortening, namespaceSuffix);
+
+        return name;
+    }
+
+    private static int calculateNameLengthWithoutNamespaceAndBlueGreenSuffix(String namespaceSuffix, int maxLengthWithSuffix) {
+        return maxLengthWithSuffix - (namespaceSuffix.length() + SystemParameters.GREEN_HOST_SUFFIX.length());
+    }
+
+    private static String correctNameSuffix(String name, String nameWithoutShortening, String namespaceSuffix) {
+        if (nameWithoutShortening.endsWith(SystemParameters.IDLE_HOST_SUFFIX)) {
+            name = placeNamespaceBeforeBlueGreenSuffix(name, namespaceSuffix, SystemParameters.IDLE_HOST_SUFFIX);
+        } else if (nameWithoutShortening.endsWith(SystemParameters.BLUE_HOST_SUFFIX)) {
+            name = placeNamespaceBeforeBlueGreenSuffix(name, namespaceSuffix, SystemParameters.BLUE_HOST_SUFFIX);
+        } else if (nameWithoutShortening.endsWith(SystemParameters.GREEN_HOST_SUFFIX)) {
+            name = placeNamespaceBeforeBlueGreenSuffix(name, namespaceSuffix, SystemParameters.GREEN_HOST_SUFFIX);
+        } else {
+            name += namespaceSuffix;
         }
 
-        return getNameWithProperLength(prefix + name, maxLength);
+        return name;
+    }
+
+    private static String placeNamespaceBeforeBlueGreenSuffix(String name, String namespaceSuffix, String blueGreenSuffix) {
```

Comment:
The problem is when the 'name' is not shortend because it is already in the range. That means that the suffix will be left there and can happen something like test-name-idle-namespace-idle. It happened during testing so I did it this way

Reply:
Does this mean that for a longer name_idle the shortening will also hash the idle part and replace it with the hash value leading the namespaced name to be 'nameHASH_namespace' without the idle in the end? Then shouldn't we extract the suffix before even shortening in the first place? That way the 'idle' or 'blue' won't get hashed in the shortening as well and will be present in a readable way and will be applied correctly.